Index: src/L2PoolInstanceWithCustomInitialize.sol
===================================================================
--- src/L2PoolInstanceWithCustomInitialize.sol
+++ src/L2PoolInstanceWithCustomInitialize.sol
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.10;
+
+import {PoolInstance} from "aave-v3-origin/contracts/instances/PoolInstance.sol";
+import {L2PoolInstance} from "aave-v3-origin/contracts/instances/L2PoolInstance.sol";
+import {Errors} from "aave-v3-origin/contracts/protocol/libraries/helpers/Errors.sol";
+import {Pool} from "aave-v3-origin/contracts/protocol/pool/Pool.sol";
+import {IPoolAddressesProvider} from "aave-v3-origin/contracts/interfaces/IPoolAddressesProvider.sol";
+import {IReserveInterestRateStrategy} from "aave-v3-origin/contracts/interfaces/IReserveInterestRateStrategy.sol";
+
+import {CustomInitialize} from "./CustomInitialize.sol";
+
+contract L2PoolInstanceWithCustomInitialize is L2PoolInstance {
+    constructor(
+        IPoolAddressesProvider provider,
+        IReserveInterestRateStrategy interestRateStrategy_
+    ) L2PoolInstance(provider, interestRateStrategy_) {}
+
+    /// @inheritdoc PoolInstance
+    function initialize(
+        IPoolAddressesProvider provider
+    ) external virtual override(Pool, PoolInstance) initializer {
+        require(
+            provider == ADDRESSES_PROVIDER,
+            Errors.InvalidAddressesProvider()
+        );
+
+        CustomInitialize._initialize(_reservesCount, _reservesList, _reserves);
+    }
+}
