===================================================================
--- lib/aave-v3-origin/src/contracts/protocol/libraries/configuration/EModeConfiguration.sol
+++ lib/aave-v3-origin-private/src/contracts/protocol/libraries/configuration/EModeConfiguration.sol
@@ -25,9 +25,9 @@
     ) internal pure returns (uint128) {
         unchecked {
             require(
                 reserveIndex < ReserveConfiguration.MAX_RESERVES_COUNT,
-                Errors.INVALID_RESERVE_INDEX
+                Errors.InvalidReserveIndex()
             );
             uint128 bit = uint128(1 << reserveIndex);
             if (enabled) {
                 return bitmap | bit;
@@ -49,9 +49,9 @@
     ) internal pure returns (bool) {
         unchecked {
             require(
                 reserveIndex < ReserveConfiguration.MAX_RESERVES_COUNT,
-                Errors.INVALID_RESERVE_INDEX
+                Errors.InvalidReserveIndex()
             );
             return (bitmap >> reserveIndex) & 1 != 0;
         }
     }
