Index: settings
===================================================================
--- settings
+++ settings
@@ -1,27 +1,38 @@
 {
   "remappings": [
-    "aave-v3-core/=src/core/",
-    "aave-v3-periphery/=src/periphery/",
-    "aave-collector-unification/=lib/aave-collector-unification/",
-    "solidity-utils/=lib/solidity-utils/src/",
-    "aave-helpers/=lib/aave-helpers/src/",
-    "aave-address-book/=lib/aave-address-book/src/",
+    "aave-address-book/=lib/aave-helpers/lib/aave-address-book/src/",
+    "aave-helpers/=lib/aave-helpers/",
+    "aave-v3-origin-tests/=lib/aave-v3-origin-private/tests/",
+    "aave-v3-origin/=lib/aave-v3-origin-private/src/",
+    "erc4626-tests/=lib/aave-helpers/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests/",
     "forge-std/=lib/forge-std/src/",
-    "ds-test/=lib/forge-std/lib/ds-test/src/",
-    "@aave/core-v2/=lib/aave-collector-unification/lib/protocol-v2/",
-    "@aave/core-v3/=lib/aave-address-book/lib/aave-v3-core/",
-    "@aave/periphery-v3/=lib/aave-address-book/lib/aave-v3-periphery/",
-    "governance-crosschain-bridges/=lib/aave-helpers/lib/governance-crosschain-bridges/",
-    "protocol-v2/=lib/aave-collector-unification/lib/protocol-v2/"
+    "openzeppelin-contracts-upgradeable/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/",
+    "openzeppelin-contracts/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/",
+    "solidity-utils/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/src/",
+    "gho-direct-minter/=lib/GhoDirectMinter/src/",
+    "lib/aave-helpers/:aave-address-book/=lib/aave-helpers/lib/aave-address-book/src/",
+    "lib/aave-helpers/:solidity-utils/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/src/",
+    "lib/aave-v3-origin-private/:solidity-utils/=lib/aave-v3-origin-private/lib/solidity-utils/src/",
+    "lib/GhoDirectMinter/:aave-address-book/=lib/aave-helpers/lib/aave-address-book/src/",
+    "lib/GhoDirectMinter/:aave-helpers/=lib/aave-helpers/",
+    "lib/GhoDirectMinter/:erc4626-tests/=lib/aave-helpers/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests/",
+    "lib/GhoDirectMinter/:forge-std/=lib/GhoDirectMinter/lib/forge-std/src/",
+    "lib/GhoDirectMinter/:solidity-utils/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/src/",
+    "@openzeppelin/contracts-upgradeable/=lib/aave-v3-origin-private/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/contracts/",
+    "@openzeppelin/contracts/=lib/aave-v3-origin-private/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/",
+    "GhoDirectMinter/=lib/GhoDirectMinter/",
+    "aave-v3-origin-private/=lib/aave-v3-origin-private/",
+    "ds-test/=lib/aave-v3-origin-private/lib/forge-std/lib/ds-test/src/",
+    "halmos-cheatcodes/=lib/aave-v3-origin-private/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/halmos-cheatcodes/src/"
   ],
   "optimizer": {
     "enabled": true,
     "runs": 200
   },
   "metadata": {
     "useLiteralContent": false,
-    "bytecodeHash": "ipfs",
+    "bytecodeHash": "none",
     "appendCBOR": true
   },
   "outputSelection": {
     "*": {
@@ -34,32 +45,30 @@
         "abi"
       ]
     }
   },
-  "evmVersion": "paris",
+  "evmVersion": "shanghai",
+  "viaIR": false,
   "libraries": {
-    "src/core/contracts/protocol/libraries/logic/BorrowLogic.sol": {
-      "BorrowLogic": "0xB64D41e260D21E9976acF420A9846aB7C76cCbF2"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/BorrowLogic.sol": {
+      "BorrowLogic": "0x5047AD5e603Ec4a2AB58aaE2321C07D8f4De6a8a"
     },
-    "src/core/contracts/protocol/libraries/logic/BridgeLogic.sol": {
-      "BridgeLogic": "0x86635418D1F55C26EEa3EDf9AD60De7D11E515Fe"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/ConfiguratorLogic.sol": {
+      "ConfiguratorLogic": "0x6E2aFD57a161d12f34f416c29619BFeAcAC8AA18"
     },
-    "src/core/contracts/protocol/libraries/logic/ConfiguratorLogic.sol": {
-      "ConfiguratorLogic": "0x0a5996586813d614f138E8E51A38435Ac65D9FaC"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/EModeLogic.sol": {
+      "EModeLogic": "0x7fcE69A2bA3e78EeB36798cde2c94C70f3A043af"
     },
-    "src/core/contracts/protocol/libraries/logic/EModeLogic.sol": {
-      "EModeLogic": "0x28DB20A0E2123a286191D267045483dA890dAea3"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/FlashLoanLogic.sol": {
+      "FlashLoanLogic": "0x4fDB5d360f946CFD25b14F346f748204c0C6a2F4"
     },
-    "src/core/contracts/protocol/libraries/logic/FlashLoanLogic.sol": {
-      "FlashLoanLogic": "0xd633446d5A5ae4480baEd3a8a370af9fbe97a9Dc"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/LiquidationLogic.sol": {
+      "LiquidationLogic": "0x5934b283f7120500253f277CCcF4521528aE34D6"
     },
-    "src/core/contracts/protocol/libraries/logic/LiquidationLogic.sol": {
-      "LiquidationLogic": "0xf36B25a326b3bB7004Af81f16c778c293974B1ad"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/PoolLogic.sol": {
+      "PoolLogic": "0x564c42578A1b270EaE16c25Da39d901245881d1F"
     },
-    "src/core/contracts/protocol/libraries/logic/PoolLogic.sol": {
-      "PoolLogic": "0xfA7d21e571EE900C7Df8880417aE50224fAF9EF9"
-    },
-    "src/core/contracts/protocol/libraries/logic/SupplyLogic.sol": {
-      "SupplyLogic": "0x7307CCE2d4DF0f3529b8E018A47dF51257271f9a"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/SupplyLogic.sol": {
+      "SupplyLogic": "0x1eF34B91afC368174F579067D1DB94325cDC7946"
     }
   }
 }
\ No newline at end of file
