===================================================================
--- @aave/core-v3/contracts/dependencies/openzeppelin/contracts/IERC20.sol
+++ lib/aave-v3-origin-private/src/contracts/dependencies/openzeppelin/contracts/IERC20.sol
@@ -1,6 +1,6 @@
-// SPDX-License-Identifier: AGPL-3.0
-pragma solidity 0.8.10;
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
 
 /**
  * @dev Interface of the ERC20 standard as defined in the EIP.
  */
