diff --git a/tmp/1868_0xD727547F49612f4590B43Bf63F0259ef38a7a488_flat/Contract.sol b/tmp/1868_0xE0a3e914836Ac28D5c42d4a278A06047c0A10c2A_flat/Contract.sol
index 4f540a8..1234c5e 100644
--- a/tmp/1868_0xD727547F49612f4590B43Bf63F0259ef38a7a488_flat/Contract.sol
+++ b/tmp/1868_0xE0a3e914836Ac28D5c42d4a278A06047c0A10c2A_flat/Contract.sol
@@ -16,7 +16,7 @@ import {AToken} from "../protocol/tokenization/AToken.sol";
  * @notice Instance of the interest bearing token for the Aave protocol
  */
 contract ATokenInstance is AToken {
-  uint256 public constant ATOKEN_REVISION = 3;
+  uint256 public constant ATOKEN_REVISION = 4;
 
   constructor(IPool pool, address rewardsController, address treasury) AToken(pool, rewardsController, treasury) {}
 
