diff --git a/tmp/1868_0x3BAe948502f2139cA25f215950bbF9d789b3821d_flat/Contract.sol b/tmp/1868_0x313B8C648Cca50704B37CCF705Dc77F7bCC5870d_flat/Contract.sol
index f6194b2..22e69af 100644
--- a/tmp/1868_0x3BAe948502f2139cA25f215950bbF9d789b3821d_flat/Contract.sol
+++ b/tmp/1868_0x313B8C648Cca50704B37CCF705Dc77F7bCC5870d_flat/Contract.sol
@@ -1,10 +1,24 @@
 // SPDX-License-Identifier: BUSL-1.1
-pragma solidity ^0.8.0;
+pragma solidity ^0.8.10;
 
-import {L2Pool} from "../protocol/pool/L2Pool.sol";
-import {IPoolAddressesProvider} from "../interfaces/IPoolAddressesProvider.sol";
-import {PoolInstance} from "./PoolInstance.sol";
+import {PoolInstance} from "aave-v3-origin/contracts/instances/PoolInstance.sol";
+import {L2PoolInstance} from "aave-v3-origin/contracts/instances/L2PoolInstance.sol";
+import {Errors} from "aave-v3-origin/contracts/protocol/libraries/helpers/Errors.sol";
+import {Pool} from "aave-v3-origin/contracts/protocol/pool/Pool.sol";
+import {IPoolAddressesProvider} from "aave-v3-origin/contracts/interfaces/IPoolAddressesProvider.sol";
+import {IReserveInterestRateStrategy} from "aave-v3-origin/contracts/interfaces/IReserveInterestRateStrategy.sol";
 
-contract L2PoolInstance is L2Pool, PoolInstance {
-  constructor(IPoolAddressesProvider provider) PoolInstance(provider) {}
+import {CustomInitialize} from "./CustomInitialize.sol";
+
+contract L2PoolInstanceWithCustomInitialize is L2PoolInstance {
+  constructor(IPoolAddressesProvider provider, IReserveInterestRateStrategy interestRateStrategy_)
+    L2PoolInstance(provider, interestRateStrategy_)
+  {}
+
+  /// @inheritdoc PoolInstance
+  function initialize(IPoolAddressesProvider provider) external virtual override(Pool, PoolInstance) initializer {
+    require(provider == ADDRESSES_PROVIDER, Errors.InvalidAddressesProvider());
+
+    CustomInitialize._initialize(_reservesCount, _reservesList, _reserves);
+  }
 }
