// @bgd-labs/aave-v3-governance-cache/dist/common/refreshCache.js
import { createRequire } from "node:module";
var __create = Object.create;
var __getProtoOf = Object.getPrototypeOf;
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __toESM = (mod, isNodeMode, target) => {
  target = mod != null ? __create(__getProtoOf(mod)) : {};
  const to = isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target;
  for (let key of __getOwnPropNames(mod))
    if (!__hasOwnProp.call(to, key))
      __defProp(to, key, {
        get: () => mod[key],
        enumerable: true
      });
  return to;
};
var __commonJS = (cb, mod) => () => (mod || cb((mod = { exports: {} }).exports, mod), mod.exports);
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {
      get: all[name],
      enumerable: true,
      configurable: true,
      set: (newValue) => all[name] = () => newValue
    });
};
var __esm = (fn, res) => () => (fn && (res = fn(fn = 0)), res);
var __require = /* @__PURE__ */ createRequire(import.meta.url);

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Ethereum.mjs
var exports_GovernanceV3Ethereum = {};
__export(exports_GovernanceV3Ethereum, {
  VOTING_STRATEGY: () => VOTING_STRATEGY,
  VOTING_PORTAL_ETH_POL: () => VOTING_PORTAL_ETH_POL,
  VOTING_PORTAL_ETH_ETH: () => VOTING_PORTAL_ETH_ETH,
  VOTING_PORTAL_ETH_AVAX: () => VOTING_PORTAL_ETH_AVAX,
  VOTING_MACHINE: () => VOTING_MACHINE,
  VM_DATA_HELPER: () => VM_DATA_HELPER,
  PC_DATA_HELPER: () => PC_DATA_HELPER,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER,
  META_DELEGATE_HELPER: () => META_DELEGATE_HELPER,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN,
  GOV_DATA_HELPER: () => GOV_DATA_HELPER,
  GOVERNANCE_POWER_STRATEGY: () => GOVERNANCE_POWER_STRATEGY,
  GOVERNANCE: () => GOVERNANCE,
  EXECUTOR_LVL_2: () => EXECUTOR_LVL_2,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_1,
  EMERGENCY_REGISTRY: () => EMERGENCY_REGISTRY,
  DATA_WAREHOUSE: () => DATA_WAREHOUSE,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER,
  CHAIN_ID: () => CHAIN_ID
});
var CROSS_CHAIN_CONTROLLER = "0xEd42a7D8559a463722Ca4beD50E0Cc05a386b0e1";
var GOVERNANCE = "0x9AEE0B04504CeF83A65AC3f0e838D0593BCb2BC7";
var PAYLOADS_CONTROLLER = "0xdAbad81aF85554E9ae636395611C58F7eC1aAEc5";
var VOTING_MACHINE = "0x617332a777780F546261247F621051d0b98975Eb";
var VOTING_PORTAL_ETH_ETH = "0xf23f7De3AC42F22eBDA17e64DC4f51FB66b8E21f";
var VOTING_PORTAL_ETH_AVAX = "0x33aCEf7365809218485873B7d0d67FeE411B5D79";
var VOTING_PORTAL_ETH_POL = "0x9b24C168d6A76b5459B1d47071a54962a4df36c3";
var PC_DATA_HELPER = "0xE3B770Dc4ae3f8bECaB3Ed12dE692c741603e16A";
var GOV_DATA_HELPER = "0x971c82c8316aD611904F95616c21ce90837f1856";
var VM_DATA_HELPER = "0x77976B51569896523EE215962Ee91ff236Fa50E8";
var META_DELEGATE_HELPER = "0x94363B11b37BC3ffe43AB09cff5A010352FE85dC";
var EMERGENCY_REGISTRY = "0x73C6Fb358dDA8e84D50e98A98F7c0dF32e15C7e9";
var GOVERNANCE_POWER_STRATEGY = "0xa198Fac58E02A5C5F8F7e877895d50cFa9ad1E04";
var GRANULAR_GUARDIAN = "0x4457cA11E90f416Cc1D3a8E1cA41C0cdEcC251d4";
var EXECUTOR_LVL_1 = "0x5300A1a15135EA4dc7aD5a167152C01EFc9b192A";
var EXECUTOR_LVL_2 = "0x17Dd33Ed0e3dD2a80E37489B8A63063161BE6957";
var VOTING_STRATEGY = "0x5642A5A5Ec284B4145563aBF319620204aCCA7f4";
var DATA_WAREHOUSE = "0x1699FE9CaDC8a0b6c93E06B62Ab4592a0fFEcF61";
var CHAIN_ID = 1;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Arbitrum.mjs
var exports_GovernanceV3Arbitrum = {};
__export(exports_GovernanceV3Arbitrum, {
  PC_DATA_HELPER: () => PC_DATA_HELPER2,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER2,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN2,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_12,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER2,
  CHAIN_ID: () => CHAIN_ID2
});
var CROSS_CHAIN_CONTROLLER2 = "0xCbFB78a3Eeaa611b826E37c80E4126c8787D29f0";
var PAYLOADS_CONTROLLER2 = "0x89644CA1bB8064760312AE4F03ea41b05dA3637C";
var PC_DATA_HELPER2 = "0xE3B770Dc4ae3f8bECaB3Ed12dE692c741603e16A";
var GRANULAR_GUARDIAN2 = "0x4922093c476CfbCF903C7C4082d2D64bAE8A37cE";
var EXECUTOR_LVL_12 = "0xFF1137243698CaA18EE364Cc966CF0e02A4e6327";
var CHAIN_ID2 = 42161;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Avalanche.mjs
var exports_GovernanceV3Avalanche = {};
__export(exports_GovernanceV3Avalanche, {
  VOTING_STRATEGY: () => VOTING_STRATEGY2,
  VOTING_MACHINE: () => VOTING_MACHINE2,
  VM_DATA_HELPER: () => VM_DATA_HELPER2,
  PC_DATA_HELPER: () => PC_DATA_HELPER3,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER3,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN3,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_13,
  DATA_WAREHOUSE: () => DATA_WAREHOUSE2,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER3,
  CL_EMERGENCY_ORACLE: () => CL_EMERGENCY_ORACLE,
  CHAIN_ID: () => CHAIN_ID3
});
var CROSS_CHAIN_CONTROLLER3 = "0x27FC7D54C893dA63C0AE6d57e1B2B13A70690928";
var CL_EMERGENCY_ORACLE = "0x41185495Bc8297a65DC46f94001DC7233775EbEe";
var VOTING_MACHINE2 = "0x9b6f5ef589A3DD08670Dd146C11C4Fb33E04494F";
var PAYLOADS_CONTROLLER3 = "0x1140CB7CAfAcC745771C2Ea31e7B5C653c5d0B80";
var PC_DATA_HELPER3 = "0xE3B770Dc4ae3f8bECaB3Ed12dE692c741603e16A";
var VM_DATA_HELPER2 = "0x77976B51569896523EE215962Ee91ff236Fa50E8";
var GRANULAR_GUARDIAN3 = "0xc1162BCb2E5E3ca4725512008c7522dF8C8B7B65";
var EXECUTOR_LVL_13 = "0x3C06dce358add17aAf230f2234bCCC4afd50d090";
var VOTING_STRATEGY2 = "0x690C218668B440204F369Af1541245d367cc805C";
var DATA_WAREHOUSE2 = "0x9626F9d60CC0B7e1c9a0A47b7f0bd618fb6f40ff";
var CHAIN_ID3 = 43114;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Optimism.mjs
var exports_GovernanceV3Optimism = {};
__export(exports_GovernanceV3Optimism, {
  PC_DATA_HELPER: () => PC_DATA_HELPER4,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER4,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN4,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_14,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER4,
  CHAIN_ID: () => CHAIN_ID4
});
var CROSS_CHAIN_CONTROLLER4 = "0x48A9FE90bce5EEd790f3F4Ce192d1C0B351fd4Ca";
var PAYLOADS_CONTROLLER4 = "0x0E1a3Af1f9cC76A62eD31eDedca291E63632e7c4";
var PC_DATA_HELPER4 = "0xE3B770Dc4ae3f8bECaB3Ed12dE692c741603e16A";
var GRANULAR_GUARDIAN4 = "0x6c5264C380C7022e54f585c4E354ffb6f221a03b";
var EXECUTOR_LVL_14 = "0x746c675dAB49Bcd5BB9Dc85161f2d7Eb435009bf";
var CHAIN_ID4 = 10;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Polygon.mjs
var exports_GovernanceV3Polygon = {};
__export(exports_GovernanceV3Polygon, {
  VOTING_STRATEGY: () => VOTING_STRATEGY3,
  VOTING_MACHINE: () => VOTING_MACHINE3,
  VM_DATA_HELPER: () => VM_DATA_HELPER3,
  PC_DATA_HELPER: () => PC_DATA_HELPER5,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER5,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN5,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_15,
  DATA_WAREHOUSE: () => DATA_WAREHOUSE3,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER5,
  CL_EMERGENCY_ORACLE: () => CL_EMERGENCY_ORACLE2,
  CHAIN_ID: () => CHAIN_ID5
});
var CROSS_CHAIN_CONTROLLER5 = "0xF6B99959F0b5e79E1CC7062E12aF632CEb18eF0d";
var CL_EMERGENCY_ORACLE2 = "0xDAFA1989A504c48Ee20a582f2891eeB25E2fA23F";
var VOTING_MACHINE3 = "0xc8a2ADC4261c6b669CdFf69E717E77C9cFeB420d";
var PAYLOADS_CONTROLLER5 = "0x401B5D0294E23637c18fcc38b1Bca814CDa2637C";
var PC_DATA_HELPER5 = "0xE3B770Dc4ae3f8bECaB3Ed12dE692c741603e16A";
var VM_DATA_HELPER3 = "0x77976B51569896523EE215962Ee91ff236Fa50E8";
var GRANULAR_GUARDIAN5 = "0x0D2CccD3dD420dC6DE2f24DB44aA22fADE290a02";
var EXECUTOR_LVL_15 = "0xDf7d0e6454DB638881302729F5ba99936EaAB233";
var VOTING_STRATEGY3 = "0x59e6CAD5d7E7b9A26a45a1d1E74C7aF008170042";
var DATA_WAREHOUSE3 = "0xf41193E25408F652AF878c47E4401A01B5E4B682";
var CHAIN_ID5 = 137;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Metis.mjs
var exports_GovernanceV3Metis = {};
__export(exports_GovernanceV3Metis, {
  PC_DATA_HELPER: () => PC_DATA_HELPER6,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER6,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN6,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_16,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER6,
  CHAIN_ID: () => CHAIN_ID6
});
var CROSS_CHAIN_CONTROLLER6 = "0x6fDaFb26915ABD6065a1E1501a37Ac438D877f70";
var PAYLOADS_CONTROLLER6 = "0x2233F8A66A728FBa6E1dC95570B25360D07D5524";
var PC_DATA_HELPER6 = "0x81d32B36380e6266e1BDd490eAC56cdB300afBe0";
var GRANULAR_GUARDIAN6 = "0x61BE97d3a0550549f67CA7421725fA73Fa2036B5";
var EXECUTOR_LVL_16 = "0x6fD45D32375d5aDB8D76275A3932c740F03a8718";
var CHAIN_ID6 = 1088;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Base.mjs
var exports_GovernanceV3Base = {};
__export(exports_GovernanceV3Base, {
  PC_DATA_HELPER: () => PC_DATA_HELPER7,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER7,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN7,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_17,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER7,
  CHAIN_ID: () => CHAIN_ID7
});
var CROSS_CHAIN_CONTROLLER7 = "0x529467C76f234F2bD359d7ecF7c660A2846b04e2";
var PAYLOADS_CONTROLLER7 = "0x2DC219E716793fb4b21548C0f009Ba3Af753ab01";
var PC_DATA_HELPER7 = "0xE3B770Dc4ae3f8bECaB3Ed12dE692c741603e16A";
var GRANULAR_GUARDIAN7 = "0xa1c6aF35E0205f42256382C05243C543FEDBf4bB";
var EXECUTOR_LVL_17 = "0x9390B1735def18560c509E2d0bc090E9d6BA257a";
var CHAIN_ID7 = 8453;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3BNB.mjs
var exports_GovernanceV3BNB = {};
__export(exports_GovernanceV3BNB, {
  PC_DATA_HELPER: () => PC_DATA_HELPER8,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER8,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN8,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_18,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER8,
  CL_EMERGENCY_ORACLE: () => CL_EMERGENCY_ORACLE3,
  CHAIN_ID: () => CHAIN_ID8
});
var CROSS_CHAIN_CONTROLLER8 = "0x9d33ee6543C9b2C8c183b8fb58fB089266cffA19";
var CL_EMERGENCY_ORACLE3 = "0xcabb46FfB38c93348Df16558DF156e9f68F9F7F1";
var PAYLOADS_CONTROLLER8 = "0xE5EF2Dd06755A97e975f7E282f828224F2C3e627";
var PC_DATA_HELPER8 = "0xE3B770Dc4ae3f8bECaB3Ed12dE692c741603e16A";
var GRANULAR_GUARDIAN8 = "0xe4FB5e3F506BE0095f38004f993D16fdA8224383";
var EXECUTOR_LVL_18 = "0x9390B1735def18560c509E2d0bc090E9d6BA257a";
var CHAIN_ID8 = 56;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Gnosis.mjs
var exports_GovernanceV3Gnosis = {};
__export(exports_GovernanceV3Gnosis, {
  PC_DATA_HELPER: () => PC_DATA_HELPER9,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER9,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN9,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_19,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER9,
  CL_EMERGENCY_ORACLE: () => CL_EMERGENCY_ORACLE4,
  CHAIN_ID: () => CHAIN_ID9
});
var CROSS_CHAIN_CONTROLLER9 = "0x8Dc5310fc9D3D7D1Bb3D1F686899c8F082316c9F";
var CL_EMERGENCY_ORACLE4 = "0xF937ffAeA1363e4Fa260760bDFA2aA8Fc911F84D";
var PAYLOADS_CONTROLLER9 = "0x9A1F491B86D09fC1484b5fab10041B189B60756b";
var PC_DATA_HELPER9 = "0xF1c11BE0b4466728DDb7991A0Ac5265646ec9672";
var GRANULAR_GUARDIAN9 = "0x4A9F571E3C1f2F13567bb59e38988e74d7d72602";
var EXECUTOR_LVL_19 = "0x1dF462e2712496373A347f8ad10802a5E95f053D";
var CHAIN_ID9 = 100;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3Scroll.mjs
var exports_GovernanceV3Scroll = {};
__export(exports_GovernanceV3Scroll, {
  PC_DATA_HELPER: () => PC_DATA_HELPER10,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER10,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN10,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_110,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER10,
  CHAIN_ID: () => CHAIN_ID10
});
var CROSS_CHAIN_CONTROLLER10 = "0x03073D3F4769f6b6604d616238fD6c636C99AD0A";
var PAYLOADS_CONTROLLER10 = "0x6b6B41c0f8C223715f712BE83ceC3c37bbfDC3fE";
var PC_DATA_HELPER10 = "0xf438e33dCCEE260ee4371F9dceF408b0d7DBe424";
var GRANULAR_GUARDIAN10 = "0xa835707d28e6C37C49d661742f2Fb5987367cEd4";
var EXECUTOR_LVL_110 = "0xc1ABF87FfAdf4908f4eC8dc54A25DCFEabAE4A24";
var CHAIN_ID10 = 534352;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3PolygonZkEvm.mjs
var exports_GovernanceV3PolygonZkEvm = {};
__export(exports_GovernanceV3PolygonZkEvm, {
  PC_DATA_HELPER: () => PC_DATA_HELPER11,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER11,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_111,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER11,
  CHAIN_ID: () => CHAIN_ID11
});
var CROSS_CHAIN_CONTROLLER11 = "0xed7e0874526B9BB9E36C7e9472ed7ed324CEeE3B";
var PAYLOADS_CONTROLLER11 = "0xa2d22795E0D85D4Cb14Cbb14Df7bb8fec1016615";
var PC_DATA_HELPER11 = "0xF1c11BE0b4466728DDb7991A0Ac5265646ec9672";
var EXECUTOR_LVL_111 = "0xe59470B3BE3293534603487E00A44C72f2CD466d";
var CHAIN_ID11 = 1101;

// node_modules/@bgd-labs/aave-address-book/dist/GovernanceV3ZkSync.mjs
var exports_GovernanceV3ZkSync = {};
__export(exports_GovernanceV3ZkSync, {
  PC_DATA_HELPER: () => PC_DATA_HELPER12,
  PAYLOADS_CONTROLLER: () => PAYLOADS_CONTROLLER12,
  GRANULAR_GUARDIAN: () => GRANULAR_GUARDIAN11,
  GOVERNANCE_GUARDIAN: () => GOVERNANCE_GUARDIAN,
  EXECUTOR_LVL_1: () => EXECUTOR_LVL_112,
  CROSS_CHAIN_CONTROLLER: () => CROSS_CHAIN_CONTROLLER12,
  CHAIN_ID: () => CHAIN_ID12
});
var CROSS_CHAIN_CONTROLLER12 = "0x800813f4714BC7A0a95310e3fB9e4f18872CA92C";
var PAYLOADS_CONTROLLER12 = "0x2E79349c3F5e4751E87b966812C9E65E805996F1";
var PC_DATA_HELPER12 = "0xe28A3235DCF1Acb8397B546bd588bAAFD7081505";
var GRANULAR_GUARDIAN11 = "0xe0e23196D42b54F262a3DE952e6B34B197D1A228";
var GOVERNANCE_GUARDIAN = "0x4257bf0746D783f0D962913d7d8AFA408B62547E";
var EXECUTOR_LVL_112 = "0x04cE39789e11a49595cD0ECEf6f4Bd54ABF4d020";
var CHAIN_ID12 = 324;

// src/common/refreshCache.ts
async function refreshCache(adapter) {
  await adapter.syncProposalCache({
    chainId: 1,
    governance: exports_GovernanceV3Ethereum.GOVERNANCE
  });
  [
    exports_GovernanceV3Ethereum,
    exports_GovernanceV3Arbitrum,
    exports_GovernanceV3Avalanche,
    exports_GovernanceV3Metis,
    exports_GovernanceV3Optimism,
    exports_GovernanceV3Gnosis,
    exports_GovernanceV3BNB,
    exports_GovernanceV3Polygon,
    exports_GovernanceV3PolygonZkEvm,
    exports_GovernanceV3Scroll,
    exports_GovernanceV3Base,
    exports_GovernanceV3ZkSync
  ].map(({ PAYLOADS_CONTROLLER: PAYLOADS_CONTROLLER13, CHAIN_ID: CHAIN_ID13 }) => {
    return adapter.syncPayloadsCache({
      chainId: CHAIN_ID13,
      payloadsController: PAYLOADS_CONTROLLER13
    });
  });
}
export {
  refreshCache
};

//# debugId=852089FDD13F425564756E2164756E21
