Index: settings
===================================================================
--- settings
+++ settings
@@ -1,30 +1,27 @@
 {
   "evmVersion": "shanghai",
   "libraries": {
-    "src/contracts/protocol/libraries/logic/BorrowLogic.sol": {
-      "BorrowLogic": "0x0B518857585421246A995a5bF04B9Ef7b9a23362"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/BorrowLogic.sol": {
+      "BorrowLogic": "0x5047AD5e603Ec4a2AB58aaE2321C07D8f4De6a8a"
     },
-    "src/contracts/protocol/libraries/logic/BridgeLogic.sol": {
-      "BridgeLogic": "0x9F7476d5E198Cec31c7AEEA73a2b2da9eD870EE6"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/ConfiguratorLogic.sol": {
+      "ConfiguratorLogic": "0x6E2aFD57a161d12f34f416c29619BFeAcAC8AA18"
     },
-    "src/contracts/protocol/libraries/logic/ConfiguratorLogic.sol": {
-      "ConfiguratorLogic": "0x4eA8C32E0AE1Cc8E8B2418f4dbc200e6cc40978B"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/EModeLogic.sol": {
+      "EModeLogic": "0x7fcE69A2bA3e78EeB36798cde2c94C70f3A043af"
     },
-    "src/contracts/protocol/libraries/logic/EModeLogic.sol": {
-      "EModeLogic": "0x0c9f369592e560444159E4C4D506CFfb379CFe99"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/FlashLoanLogic.sol": {
+      "FlashLoanLogic": "0x4fDB5d360f946CFD25b14F346f748204c0C6a2F4"
     },
-    "src/contracts/protocol/libraries/logic/FlashLoanLogic.sol": {
-      "FlashLoanLogic": "0x0c1b3C6643b5505f552a726864f10e6e44a9A455"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/LiquidationLogic.sol": {
+      "LiquidationLogic": "0x5934b283f7120500253f277CCcF4521528aE34D6"
     },
-    "src/contracts/protocol/libraries/logic/LiquidationLogic.sol": {
-      "LiquidationLogic": "0xdD87Bce757Ea6b73436761c5D44F182471b8A100"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/PoolLogic.sol": {
+      "PoolLogic": "0x564c42578A1b270EaE16c25Da39d901245881d1F"
     },
-    "src/contracts/protocol/libraries/logic/PoolLogic.sol": {
-      "PoolLogic": "0xeAD5169A6eb93e0EFCB34Ec1F16624e305e08EBC"
-    },
-    "src/contracts/protocol/libraries/logic/SupplyLogic.sol": {
-      "SupplyLogic": "0x231cd9e4fc7A940b33f201af67C9d50091E51CD7"
+    "lib/aave-v3-origin-private/src/contracts/protocol/libraries/logic/SupplyLogic.sol": {
+      "SupplyLogic": "0x1eF34B91afC368174F579067D1DB94325cDC7946"
     }
   },
   "metadata": {
     "appendCBOR": true,
@@ -36,31 +33,40 @@
     "runs": 200
   },
   "outputSelection": {
     "*": {
+      "": [
+        "*"
+      ],
       "*": [
-        "abi",
-        "evm.bytecode.object",
-        "evm.bytecode.sourceMap",
-        "evm.bytecode.linkReferences",
-        "evm.deployedBytecode.object",
-        "evm.deployedBytecode.sourceMap",
-        "evm.deployedBytecode.linkReferences",
-        "evm.deployedBytecode.immutableReferences",
-        "evm.methodIdentifiers",
-        "metadata"
+        "*"
       ]
     }
   },
   "remappings": [
-    "solidity-utils/=lib/solidity-utils/src/",
+    "aave-address-book/=lib/aave-helpers/lib/aave-address-book/src/",
+    "aave-helpers/=lib/aave-helpers/",
+    "aave-v3-origin-tests/=lib/aave-v3-origin-private/tests/",
+    "aave-v3-origin/=lib/aave-v3-origin-private/src/",
+    "erc4626-tests/=lib/aave-helpers/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests/",
     "forge-std/=lib/forge-std/src/",
-    "ds-test/=lib/forge-std/lib/ds-test/src/",
-    "openzeppelin-contracts-upgradeable/=lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/",
-    "openzeppelin-contracts/=lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/",
-    "@openzeppelin/contracts-upgradeable/=lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/contracts/",
-    "@openzeppelin/contracts/=lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/",
-    "erc4626-tests/=lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests/",
-    "halmos-cheatcodes/=lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/halmos-cheatcodes/src/"
+    "openzeppelin-contracts-upgradeable/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/",
+    "openzeppelin-contracts/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/",
+    "solidity-utils/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/src/",
+    "gho-direct-minter/=lib/GhoDirectMinter/src/",
+    "lib/aave-helpers/:aave-address-book/=lib/aave-helpers/lib/aave-address-book/src/",
+    "lib/aave-helpers/:solidity-utils/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/src/",
+    "lib/aave-v3-origin-private/:solidity-utils/=lib/aave-v3-origin-private/lib/solidity-utils/src/",
+    "lib/GhoDirectMinter/:aave-address-book/=lib/aave-helpers/lib/aave-address-book/src/",
+    "lib/GhoDirectMinter/:aave-helpers/=lib/aave-helpers/",
+    "lib/GhoDirectMinter/:erc4626-tests/=lib/aave-helpers/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/erc4626-tests/",
+    "lib/GhoDirectMinter/:forge-std/=lib/GhoDirectMinter/lib/forge-std/src/",
+    "lib/GhoDirectMinter/:solidity-utils/=lib/aave-helpers/lib/aave-address-book/lib/aave-v3-origin/lib/solidity-utils/src/",
+    "@openzeppelin/contracts-upgradeable/=lib/aave-v3-origin-private/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/contracts/",
+    "@openzeppelin/contracts/=lib/aave-v3-origin-private/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/",
+    "GhoDirectMinter/=lib/GhoDirectMinter/",
+    "aave-v3-origin-private/=lib/aave-v3-origin-private/",
+    "ds-test/=lib/aave-v3-origin-private/lib/forge-std/lib/ds-test/src/",
+    "halmos-cheatcodes/=lib/aave-v3-origin-private/lib/solidity-utils/lib/openzeppelin-contracts-upgradeable/lib/halmos-cheatcodes/src/"
   ],
   "viaIR": false
 }
\ No newline at end of file
