// @bgd-labs/aave-address-book/dist/AaveV3Base.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ts/AaveV3Base.ts
var AaveV3Base_exports = {};
__export(AaveV3Base_exports, {
  AAVE_PROTOCOL_DATA_PROVIDER: () => AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN: () => ACL_ADMIN,
  ACL_MANAGER: () => ACL_MANAGER,
  ASSETS: () => ASSETS,
  CAPS_PLUS_RISK_STEWARD: () => CAPS_PLUS_RISK_STEWARD,
  CHAIN_ID: () => CHAIN_ID,
  COLLECTOR: () => COLLECTOR,
  CONFIG_ENGINE: () => CONFIG_ENGINE,
  DEBT_SWAP_ADAPTER: () => DEBT_SWAP_ADAPTER,
  DEFAULT_A_TOKEN_IMPL_REV_1: () => DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER: () => DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1: () => DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  EMISSION_MANAGER: () => EMISSION_MANAGER,
  EXTERNAL_LIBRARIES: () => EXTERNAL_LIBRARIES,
  E_MODES: () => E_MODES,
  FREEZING_STEWARD: () => FREEZING_STEWARD,
  L2_ENCODER: () => L2_ENCODER,
  ORACLE: () => ORACLE,
  POOL: () => POOL,
  POOL_ADDRESSES_PROVIDER: () => POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY: () => POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR: () => POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL: () => POOL_CONFIGURATOR_IMPL,
  POOL_IMPL: () => POOL_IMPL,
  PRICE_ORACLE_SENTINEL: () => PRICE_ORACLE_SENTINEL,
  REPAY_WITH_COLLATERAL_ADAPTER: () => REPAY_WITH_COLLATERAL_ADAPTER,
  RISK_STEWARD: () => RISK_STEWARD,
  STATIC_A_TOKEN_FACTORY: () => STATIC_A_TOKEN_FACTORY,
  STATIC_A_TOKEN_ORACLE: () => STATIC_A_TOKEN_ORACLE,
  SWAP_COLLATERAL_ADAPTER: () => SWAP_COLLATERAL_ADAPTER,
  UI_INCENTIVE_DATA_PROVIDER: () => UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER: () => UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER: () => WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY: () => WETH_GATEWAY,
  WITHDRAW_SWAP_ADAPTER: () => WITHDRAW_SWAP_ADAPTER
});
module.exports = __toCommonJS(AaveV3Base_exports);
var POOL_ADDRESSES_PROVIDER = "0xe20fCBdBfFC4Dd138cE8b2E6FBb6CB49777ad64D";
var POOL = "0xA238Dd80C259a72e81d7e4664a9801593F98d1c5";
var POOL_CONFIGURATOR = "0x5731a04B1E775f0fdd454Bf70f3335886e9A96be";
var ORACLE = "0x2Cc0Fc26eD4563A5ce5e8bdcfe1A2878676Ae156";
var PRICE_ORACLE_SENTINEL = "0x943AcD0c93d7a8Bee7dA5Fd0DC3d0028237074d6";
var ACL_ADMIN = "0x9390B1735def18560c509E2d0bc090E9d6BA257a";
var ACL_MANAGER = "0x43955b0899Ab7232E3a454cf84AedD22Ad46FD33";
var AAVE_PROTOCOL_DATA_PROVIDER = "0xd82a47fdebB5bf5329b09441C3DaB4b5df2153Ad";
var POOL_IMPL = "0x3a9C471F13C9cA1eBDf440cf713C8404e498F9C3";
var POOL_CONFIGURATOR_IMPL = "0x4816b2C2895f97fB918f1aE7Da403750a0eE372e";
var DEFAULT_INCENTIVES_CONTROLLER = "0xf9cc4F0D883F1a1eb2c253bdb46c254Ca51E1F44";
var EMISSION_MANAGER = "0x6533A273F3aC84Df91DCD654D6EBAbA73687e246";
var COLLECTOR = "0xBA9424d650A4F5c80a0dA641254d1AcCE2A37057";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0x98F409Fc4A42F34AE3c326c7f48ED01ae8cAeC69";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0x2425A746911128c2eAA7bEBDc9Bc452eE52208a1";
var CAPS_PLUS_RISK_STEWARD = "0x12DEB4025b79f2B43f6aeF079F9D77C3f9a67bb6";
var RISK_STEWARD = "0xD3DE4b3571744EB77946d65aBF01408902E92c4E";
var FREEZING_STEWARD = "0x4A4c73d563395ad827511F70097d4Ef82E653805";
var DEBT_SWAP_ADAPTER = "0xb12e82DF057BF16ecFa89D7D089dc7E5C1Dc057B";
var WITHDRAW_SWAP_ADAPTER = "0x5598BbFA2f4fE8151f45bBA0a3edE1b54B51a0a9";
var SWAP_COLLATERAL_ADAPTER = "0x2E549104c516b8657A7D888494DfbAbD7C70b464";
var REPAY_WITH_COLLATERAL_ADAPTER = "0x63dfa7c09Dc2Ff4030d6B8Dc2ce6262BF898C8A4";
var STATIC_A_TOKEN_FACTORY = "0x940F9a5d5F9ED264990D0eaee1F3DD60B4Cb9A22";
var STATIC_A_TOKEN_ORACLE = "0xe63eAf6DAb1045689BD3a332bC596FfcF54A5C88";
var CONFIG_ENGINE = "0xD828E29E2268f56309dDbbAB697ae85DCEb43B9d";
var L2_ENCODER = "0x39e97c588B2907Fb67F44fea256Ae3BA064207C5";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0x2f6571d3Eb9a4e350C68C36bCD2afe39530078E2";
var UI_INCENTIVE_DATA_PROVIDER = "0x91E04cf78e53aEBe609e8a7f2003e7EECD743F2B";
var UI_POOL_DATA_PROVIDER = "0x68100bD5345eA474D93577127C11F39FF8463e93";
var WALLET_BALANCE_PROVIDER = "0x5779b29B0a34577d927E8D511B595ef9abbFAE82";
var WETH_GATEWAY = "0x729b3EA8C005AbC58c9150fb57Ec161296F06766";
var CHAIN_ID = 8453;
var ASSETS = {
  WETH: {
    decimals: 18,
    id: 0,
    UNDERLYING: "0x4200000000000000000000000000000000000006",
    A_TOKEN: "0xD4a0e0b9149BCee3C920d2E00b5dE09138fd8bb7",
    V_TOKEN: "0x24e6e0795b3c7c71D965fCc4f371803d1c1DcA1E",
    INTEREST_RATE_STRATEGY: "0x86AB1C62A8bf868E1b3E1ab87d587Aba6fbCbDC5",
    ORACLE: "0x71041dddad3595F9CEd3DcCFBe3D1F4b0a16Bb70",
    STATIC_A_TOKEN: "0x468973e3264F2aEba0417A8f2cD0Ec397E738898"
  },
  cbETH: {
    decimals: 18,
    id: 1,
    UNDERLYING: "0x2Ae3F1Ec7F1F5012CFEab0185bfc7aa3cf0DEc22",
    A_TOKEN: "0xcf3D55c10DB69f28fD1A75Bd73f3D8A2d9c595ad",
    V_TOKEN: "0x1DabC36f19909425f654777249815c073E8Fd79F",
    INTEREST_RATE_STRATEGY: "0x86AB1C62A8bf868E1b3E1ab87d587Aba6fbCbDC5",
    ORACLE: "0x8e11Ad4531826ff47BD8157a2c705F5422Da6A61",
    STATIC_A_TOKEN: "0x16A004065dfb11276DcB29Dc03fb8A85f9A43C6e"
  },
  USDbC: {
    decimals: 6,
    id: 2,
    UNDERLYING: "0xd9aAEc86B65D86f6A7B5B1b0c42FFA531710b6CA",
    A_TOKEN: "0x0a1d576f3eFeF75b330424287a95A366e8281D54",
    V_TOKEN: "0x7376b2F323dC56fCd4C191B34163ac8a84702DAB",
    INTEREST_RATE_STRATEGY: "0x86AB1C62A8bf868E1b3E1ab87d587Aba6fbCbDC5",
    ORACLE: "0x978D8878b53Fbe40dab7D4AB47b97AB622FFeF9f",
    STATIC_A_TOKEN: "0x6fCe2756794128B1771324caA860965801DCbCdB"
  },
  wstETH: {
    decimals: 18,
    id: 3,
    UNDERLYING: "0xc1CBa3fCea344f92D9239c08C0568f6F2F0ee452",
    A_TOKEN: "0x99CBC45ea5bb7eF3a5BC08FB1B7E56bB2442Ef0D",
    V_TOKEN: "0x41A7C3f5904ad176dACbb1D99101F59ef0811DC1",
    INTEREST_RATE_STRATEGY: "0x86AB1C62A8bf868E1b3E1ab87d587Aba6fbCbDC5",
    ORACLE: "0x56038D3998C42db18ba3B821bD1EbaB9B678e657",
    STATIC_A_TOKEN: "0x03916e49f794Ab877eFA23597627eE8094E6cbB0"
  },
  USDC: {
    decimals: 6,
    id: 4,
    UNDERLYING: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    A_TOKEN: "0x4e65fE4DbA92790696d040ac24Aa414708F5c0AB",
    V_TOKEN: "0x59dca05b6c26dbd64b5381374aAaC5CD05644C28",
    INTEREST_RATE_STRATEGY: "0x86AB1C62A8bf868E1b3E1ab87d587Aba6fbCbDC5",
    ORACLE: "0x978D8878b53Fbe40dab7D4AB47b97AB622FFeF9f",
    STATIC_A_TOKEN: "0x4EA71A20e655794051D1eE8b6e4A3269B13ccaCc"
  },
  weETH: {
    decimals: 18,
    id: 5,
    UNDERLYING: "0x04C0599Ae5A44757c0af6F9eC3b93da8976c150A",
    A_TOKEN: "0x7C307e128efA31F540F2E2d976C995E0B65F51F6",
    V_TOKEN: "0x8D2e3F1f4b38AA9f1ceD22ac06019c7561B03901",
    INTEREST_RATE_STRATEGY: "0x86AB1C62A8bf868E1b3E1ab87d587Aba6fbCbDC5",
    ORACLE: "0xFc4d1d7a8FD1E6719e361e16044b460737F12C44",
    STATIC_A_TOKEN: "0x588159E0d360ffAA978330812f9234818ab46E8E"
  },
  cbBTC: {
    decimals: 8,
    id: 6,
    UNDERLYING: "0xcbB7C0000aB88B473b1f5aFd9ef808440eed33Bf",
    A_TOKEN: "0xBdb9300b7CDE636d9cD4AFF00f6F009fFBBc8EE6",
    V_TOKEN: "0x05e08702028de6AaD395DC6478b554a56920b9AD",
    INTEREST_RATE_STRATEGY: "0x86AB1C62A8bf868E1b3E1ab87d587Aba6fbCbDC5",
    ORACLE: "0x64c911996D3c6aC71f9b455B1E8E7266BcbD848F"
  }
};
var E_MODES = {
  "1": {
    label: "ETH correlated",
    collateralBitmap: "43",
    collateralAssets: [
      "0x4200000000000000000000000000000000000006",
      "0x2Ae3F1Ec7F1F5012CFEab0185bfc7aa3cf0DEc22",
      "0xc1CBa3fCea344f92D9239c08C0568f6F2F0ee452",
      "0x04C0599Ae5A44757c0af6F9eC3b93da8976c150A"
    ],
    borrowableBitmap: "43",
    borrowableAssets: [
      "0x4200000000000000000000000000000000000006",
      "0x2Ae3F1Ec7F1F5012CFEab0185bfc7aa3cf0DEc22",
      "0xc1CBa3fCea344f92D9239c08C0568f6F2F0ee452",
      "0x04C0599Ae5A44757c0af6F9eC3b93da8976c150A"
    ],
    ltv: 9e3,
    liquidationThreshold: 9300,
    liquidationBonus: 10200
  }
};
var EXTERNAL_LIBRARIES = {};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CAPS_PLUS_RISK_STEWARD,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEBT_SWAP_ADAPTER,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  FREEZING_STEWARD,
  L2_ENCODER,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  PRICE_ORACLE_SENTINEL,
  REPAY_WITH_COLLATERAL_ADAPTER,
  RISK_STEWARD,
  STATIC_A_TOKEN_FACTORY,
  STATIC_A_TOKEN_ORACLE,
  SWAP_COLLATERAL_ADAPTER,
  UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY,
  WITHDRAW_SWAP_ADAPTER
});
