// @bgd-labs/aave-address-book/dist/AaveV3Scroll.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ts/AaveV3Scroll.ts
var AaveV3Scroll_exports = {};
__export(AaveV3Scroll_exports, {
  AAVE_PROTOCOL_DATA_PROVIDER: () => AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN: () => ACL_ADMIN,
  ACL_MANAGER: () => ACL_MANAGER,
  ASSETS: () => ASSETS,
  CAPS_PLUS_RISK_STEWARD: () => CAPS_PLUS_RISK_STEWARD,
  CHAIN_ID: () => CHAIN_ID,
  COLLECTOR: () => COLLECTOR,
  CONFIG_ENGINE: () => CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1: () => DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER: () => DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1: () => DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  EMISSION_MANAGER: () => EMISSION_MANAGER,
  EXTERNAL_LIBRARIES: () => EXTERNAL_LIBRARIES,
  E_MODES: () => E_MODES,
  FREEZING_STEWARD: () => FREEZING_STEWARD,
  L2_ENCODER: () => L2_ENCODER,
  ORACLE: () => ORACLE,
  POOL: () => POOL,
  POOL_ADDRESSES_PROVIDER: () => POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY: () => POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR: () => POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL: () => POOL_CONFIGURATOR_IMPL,
  POOL_IMPL: () => POOL_IMPL,
  PRICE_ORACLE_SENTINEL: () => PRICE_ORACLE_SENTINEL,
  RISK_STEWARD: () => RISK_STEWARD,
  STATIC_A_TOKEN_FACTORY: () => STATIC_A_TOKEN_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER: () => UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER: () => UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER: () => WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY: () => WETH_GATEWAY
});
module.exports = __toCommonJS(AaveV3Scroll_exports);
var POOL_ADDRESSES_PROVIDER = "0x69850D0B276776781C063771b161bd8894BCdD04";
var POOL = "0x11fCfe756c05AD438e312a7fd934381537D3cFfe";
var POOL_CONFIGURATOR = "0x32BCab42a2bb5AC577D24b425D46d8b8e0Df9b7f";
var ORACLE = "0x04421D8C506E2fA2371a08EfAaBf791F624054F3";
var PRICE_ORACLE_SENTINEL = "0xfD0Ba55775C1e53f50736FA5528d8aa45FBcA391";
var ACL_ADMIN = "0xc1ABF87FfAdf4908f4eC8dc54A25DCFEabAE4A24";
var ACL_MANAGER = "0x7633F981D87dC6307227de9383D2ce7243158081";
var AAVE_PROTOCOL_DATA_PROVIDER = "0xe2108b60623C6Dcf7bBd535bD15a451fd0811f7b";
var POOL_IMPL = "0xcB2107Ace932591C57EB5D07A135F1F3Dd613Dc0";
var POOL_CONFIGURATOR_IMPL = "0x4816b2C2895f97fB918f1aE7Da403750a0eE372e";
var DEFAULT_INCENTIVES_CONTROLLER = "0xa3f3100C4f1D0624DB9DB97b40C13885Ce297799";
var EMISSION_MANAGER = "0x6091546836DAe0487A50E300da3F237727441D90";
var COLLECTOR = "0x90eB541e1a431D8a30ED85A77675D1F001128cb5";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0x92EDe4ABd9df4Bfb49b4d723e4c932e35c47C54C";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0x49bA16C08130FF8cFADE263B49387A8555bc057B";
var L2_ENCODER = "0x8714E5ED2d8edD4E88eFf66637C3FE8eCf2B8C40";
var CAPS_PLUS_RISK_STEWARD = "0xc4dcA550EF04FD0f8AbD4c384575Fb8a8123054e";
var RISK_STEWARD = "0x64093fe5f8Cf62aFb4377cf7EF4373537fe9155B";
var FREEZING_STEWARD = "0xc68D0C511076283075bD5cc9aA61E43673135f37";
var CONFIG_ENGINE = "0xC0FE411b2606a06f64C1536E8F92FA6a9d5a89B8";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0xFBedc64AeE24921cb43004312B9eF367a4162b57";
var UI_INCENTIVE_DATA_PROVIDER = "0x5d4D4007A4c6336550DdAa2a7c0d5e7972eebd16";
var UI_POOL_DATA_PROVIDER = "0x5598BbFA2f4fE8151f45bBA0a3edE1b54B51a0a9";
var WALLET_BALANCE_PROVIDER = "0xE51642875Af07Fea1B03d821E41a9DB24bC4447E";
var WETH_GATEWAY = "0x31A239f3e39c5D8BA6B201bA81ed584492Ae960F";
var STATIC_A_TOKEN_FACTORY = "0x970b77b96D94966939a8F867c2BfC2aC0310C0e7";
var CHAIN_ID = 534352;
var ASSETS = {
  WETH: {
    decimals: 18,
    id: 0,
    UNDERLYING: "0x5300000000000000000000000000000000000004",
    A_TOKEN: "0xf301805bE1Df81102C957f6d4Ce29d2B8c056B2a",
    V_TOKEN: "0xfD7344CeB1Df9Cf238EcD667f4A6F99c6Ef44a56",
    INTEREST_RATE_STRATEGY: "0xC37353E5766164D8654D3CB395acfDcA4c2E7Ddc",
    ORACLE: "0x6bF14CB0A831078629D993FDeBcB182b21A8774C",
    STATIC_A_TOKEN: "0x6b9DfaC194fa78a1882680E2cE19194D006AeEfd"
  },
  USDC: {
    decimals: 6,
    id: 1,
    UNDERLYING: "0x06eFdBFf2a14a7c8E15944D1F4A48F9F95F663A4",
    A_TOKEN: "0x1D738a3436A8C49CefFbaB7fbF04B660fb528CbD",
    V_TOKEN: "0x3d2E209af5BFa79297C88D6b57F89d792F6E28EE",
    INTEREST_RATE_STRATEGY: "0xC37353E5766164D8654D3CB395acfDcA4c2E7Ddc",
    ORACLE: "0x427Fd98dbD1DbC2D4e792350caBe7c9665F35bee",
    STATIC_A_TOKEN: "0x9fA123bC7E6b61cC8a9D893673a4C6E5392FF4A7"
  },
  wstETH: {
    decimals: 18,
    id: 2,
    UNDERLYING: "0xf610A9dfB7C89644979b4A0f27063E9e7d7Cda32",
    A_TOKEN: "0x5B1322eeb46240b02e20062b8F0F9908d525B09c",
    V_TOKEN: "0x8a035644322129800C3f747f54Db0F4d3c0A2877",
    INTEREST_RATE_STRATEGY: "0xC37353E5766164D8654D3CB395acfDcA4c2E7Ddc",
    ORACLE: "0x4EdAbf45e78363b8Dcd763bBbd05665c6e24975C",
    STATIC_A_TOKEN: "0x6e368c4dBf083e18a29aE63FC06AF9deDb3242F0"
  },
  weETH: {
    decimals: 18,
    id: 3,
    UNDERLYING: "0x01f0a31698C4d065659b9bdC21B3610292a1c506",
    A_TOKEN: "0xd80A5e16DBDC52Bd1C947CEDfA22c562Be9129C8",
    V_TOKEN: "0x009D88C6a6B4CaA240b71C98BA93732e26F2A55A",
    INTEREST_RATE_STRATEGY: "0xC37353E5766164D8654D3CB395acfDcA4c2E7Ddc",
    ORACLE: "0x32f924C0e0F1Abf5D1ff35B05eBc5E844dEdD2A9"
  }
};
var E_MODES = {
  "1": {
    label: "ETH correlated",
    collateralBitmap: "13",
    collateralAssets: [
      "0x5300000000000000000000000000000000000004",
      "0xf610A9dfB7C89644979b4A0f27063E9e7d7Cda32",
      "0x01f0a31698C4d065659b9bdC21B3610292a1c506"
    ],
    borrowableBitmap: "13",
    borrowableAssets: [
      "0x5300000000000000000000000000000000000004",
      "0xf610A9dfB7C89644979b4A0f27063E9e7d7Cda32",
      "0x01f0a31698C4d065659b9bdC21B3610292a1c506"
    ],
    ltv: 9e3,
    liquidationThreshold: 9300,
    liquidationBonus: 10100
  }
};
var EXTERNAL_LIBRARIES = {};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CAPS_PLUS_RISK_STEWARD,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  FREEZING_STEWARD,
  L2_ENCODER,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  PRICE_ORACLE_SENTINEL,
  RISK_STEWARD,
  STATIC_A_TOKEN_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY
});
