// @bgd-labs/aave-address-book/dist/AaveV3EthereumLido.mjs
// src/ts/AaveV3EthereumLido.ts
var POOL_ADDRESSES_PROVIDER = "0xcfBf336fe147D643B9Cb705648500e101504B16d";
var POOL = "0x4e033931ad43597d96D6bcc25c280717730B58B1";
var POOL_CONFIGURATOR = "0x342631c6CeFC9cfbf97b2fe4aa242a236e1fd517";
var ORACLE = "0xE3C061981870C0C7b1f3C4F4bB36B95f1F260BE6";
var ACL_ADMIN = "0x5300A1a15135EA4dc7aD5a167152C01EFc9b192A";
var ACL_MANAGER = "0x013E2C7567b6231e865BB9273F8c7656103611c0";
var AAVE_PROTOCOL_DATA_PROVIDER = "0x08795CFE08C7a81dCDFf482BbAAF474B240f31cD";
var POOL_IMPL = "0xf5b4664CB6D13189345119c60a948cdc7785d0FE";
var POOL_CONFIGURATOR_IMPL = "0x4816b2C2895f97fB918f1aE7Da403750a0eE372e";
var DEFAULT_INCENTIVES_CONTROLLER = "0x8164Cc65827dcFe994AB23944CBC90e0aa80bFcb";
var EMISSION_MANAGER = "0x223d844fc4B006D67c0cDbd39371A9F73f69d974";
var COLLECTOR = "0x464C71f6c2F760DdA6093dCB91C24c39e5d6e18c";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0x7F8Fc14D462bdF93c681c1f2Fd615389bF969Fb2";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0x3E59212c34588a63350142EFad594a20C88C2CEd";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0xC6cAB8D39D93DC0Bd5986E7Ce5Bb956E30103A43";
var REPAY_WITH_COLLATERAL_ADAPTER = "0x66E1aBdb06e7363a618D65a910c540dfED23754f";
var SWAP_COLLATERAL_ADAPTER = "0xD0887AA7fEBC8962c622493646195e7c76D94fCE";
var WITHDRAW_SWAP_ADAPTER = "0x43eDB797834151D041619EEF833Edc784B509dAE";
var UI_INCENTIVE_DATA_PROVIDER = "0xe3dFf4052F0bF6134ACb73bEaE8fe2317d71F047";
var UI_POOL_DATA_PROVIDER = "0x3F78BBD206e4D3c504Eb854232EdA7e47E9Fd8FC";
var WALLET_BALANCE_PROVIDER = "0xC7be5307ba715ce89b152f3Df0658295b3dbA8E2";
var WETH_GATEWAY = "0x0B8C700917a6991FEa7198dDFC80bc8962b5055D";
var CONFIG_ENGINE = "0x1097eDb85392932b7dCB630baDDC8A6D73585218";
var CAPS_PLUS_RISK_STEWARD = "0x3843b29118fFC18d5d12EE079d0324E1bF115e69";
var RISK_STEWARD = "0x30adC2f98ff78fDde12F191Acb82699f640694FB";
var DEBT_SWAP_ADAPTER = "0xd1B2dec98A95B773C4909B5CD8FB455F467A527f";
var CHAIN_ID = 1;
var ASSETS = {
  wstETH: {
    decimals: 18,
    id: 0,
    UNDERLYING: "0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0",
    A_TOKEN: "0xC035a7cf15375cE2706766804551791aD035E0C2",
    V_TOKEN: "0xE439edd2625772AA635B437C099C607B6eb7d35f",
    INTEREST_RATE_STRATEGY: "0x8958b1C39269167527821f8c276Ef7504883f2fa",
    ORACLE: "0xB4aB0c94159bc2d8C133946E7241368fc2F2a010"
  },
  WETH: {
    decimals: 18,
    id: 1,
    UNDERLYING: "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    A_TOKEN: "0xfA1fDbBD71B0aA16162D76914d69cD8CB3Ef92da",
    V_TOKEN: "0x91b7d78BF92db564221f6B5AeE744D1727d1Dd1e",
    INTEREST_RATE_STRATEGY: "0x8958b1C39269167527821f8c276Ef7504883f2fa",
    ORACLE: "0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419"
  },
  USDS: {
    decimals: 18,
    id: 2,
    UNDERLYING: "0xdC035D45d973E3EC169d2276DDab16f1e407384F",
    A_TOKEN: "0x09AA30b182488f769a9824F15E6Ce58591Da4781",
    V_TOKEN: "0x2D9fe18b6c35FE439cC15D932cc5C943bf2d901E",
    INTEREST_RATE_STRATEGY: "0x8958b1C39269167527821f8c276Ef7504883f2fa",
    ORACLE: "0x4F01b76391A05d32B20FA2d05dD5963eE8db20E6"
  },
  USDC: {
    decimals: 6,
    id: 3,
    UNDERLYING: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    A_TOKEN: "0x2A1FBcb52Ed4d9b23daD17E1E8Aed4BB0E6079b8",
    V_TOKEN: "0xeD90dE2D824Ee766c6Fd22E90b12e598f681dc9F",
    INTEREST_RATE_STRATEGY: "0x8958b1C39269167527821f8c276Ef7504883f2fa",
    ORACLE: "0x736bF902680e68989886e9807CD7Db4B3E015d3C"
  },
  ezETH: {
    decimals: 18,
    id: 4,
    UNDERLYING: "0xbf5495Efe5DB9ce00f80364C8B423567e58d2110",
    A_TOKEN: "0x74e5664394998f13B07aF42446380ACef637969f",
    V_TOKEN: "0x08e1bba76D27841dD91FAb4b3a636A0D5CF8c3E9",
    INTEREST_RATE_STRATEGY: "0x8958b1C39269167527821f8c276Ef7504883f2fa",
    ORACLE: "0x68C9c7Bf43DBd0EBab102116bc7C3C9f7d9297Ee"
  },
  sUSDe: {
    decimals: 18,
    id: 5,
    UNDERLYING: "0x9D39A5DE30e57443BfF2A8307A4256c8797A3497",
    A_TOKEN: "0xc2015641564a5914A17CB9A92eC8d8feCfa8f2D0",
    V_TOKEN: "0x2ABbAab3EF4e4A899d39e7EC996b5715E76b399a",
    INTEREST_RATE_STRATEGY: "0x8958b1C39269167527821f8c276Ef7504883f2fa",
    ORACLE: "0xb37aE8aBa6C0C1Bf2c509fc06E11aa4AF29B665A"
  }
};
var E_MODES = {
  "1": {
    label: "ETH correlated",
    collateralBitmap: "3",
    collateralAssets: [
      "0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0",
      "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"
    ],
    borrowableBitmap: "2",
    borrowableAssets: ["0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"],
    ltv: 9350,
    liquidationThreshold: 9550,
    liquidationBonus: 10100
  },
  "2": {
    label: "LRT Stablecoins main",
    collateralBitmap: "16",
    collateralAssets: ["0xbf5495Efe5DB9ce00f80364C8B423567e58d2110"],
    borrowableBitmap: "4",
    borrowableAssets: ["0xdC035D45d973E3EC169d2276DDab16f1e407384F"],
    ltv: 7500,
    liquidationThreshold: 7800,
    liquidationBonus: 10750
  },
  "3": {
    label: "LRT wstETH main",
    collateralBitmap: "16",
    collateralAssets: ["0xbf5495Efe5DB9ce00f80364C8B423567e58d2110"],
    borrowableBitmap: "1",
    borrowableAssets: ["0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0"],
    ltv: 9300,
    liquidationThreshold: 9500,
    liquidationBonus: 10100
  },
  "4": {
    label: "sUSDe Stablecoins",
    collateralBitmap: "32",
    collateralAssets: ["0x9D39A5DE30e57443BfF2A8307A4256c8797A3497"],
    borrowableBitmap: "12",
    borrowableAssets: [
      "0xdC035D45d973E3EC169d2276DDab16f1e407384F",
      "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"
    ],
    ltv: 9e3,
    liquidationThreshold: 9200,
    liquidationBonus: 10300
  }
};
var EXTERNAL_LIBRARIES = {};
export {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CAPS_PLUS_RISK_STEWARD,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEBT_SWAP_ADAPTER,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  REPAY_WITH_COLLATERAL_ADAPTER,
  RISK_STEWARD,
  SWAP_COLLATERAL_ADAPTER,
  UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY,
  WITHDRAW_SWAP_ADAPTER
};
