// @bgd-labs/aave-address-book/dist/AaveV3Gnosis.mjs
// src/ts/AaveV3Gnosis.ts
var POOL_ADDRESSES_PROVIDER = "0x36616cf17557639614c1cdDb356b1B83fc0B2132";
var POOL = "0xb50201558B00496A145fE76f7424749556E326D8";
var POOL_CONFIGURATOR = "0x7304979ec9E4EaA0273b6A037a31c4e9e5A75D16";
var ORACLE = "0xeb0a051be10228213BAEb449db63719d6742F7c4";
var ACL_ADMIN = "0x1dF462e2712496373A347f8ad10802a5E95f053D";
var ACL_MANAGER = "0xEc710f59005f48703908bC519D552Df5B8472614";
var AAVE_PROTOCOL_DATA_PROVIDER = "0x57038C3e3Fe0a170BB72DE2fD56E98e4d1a69717";
var POOL_IMPL = "0xF2C312BfAF4CF0429DB4DA15a0cf5F770Ea3E770";
var POOL_CONFIGURATOR_IMPL = "0x4816b2C2895f97fB918f1aE7Da403750a0eE372e";
var DEFAULT_INCENTIVES_CONTROLLER = "0xaD4F91D26254B6B0C6346b390dDA2991FDE2F20d";
var EMISSION_MANAGER = "0x41585C50524fb8c3899B43D7D797d9486AAc94DB";
var COLLECTOR = "0x3e652E97ff339B73421f824F5b03d75b62F1Fb51";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0x589750BA8aF186cE5B55391B0b7148cAD43a1619";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0xBeC519531F0E78BcDdB295242fA4EC5251B38574";
var CONFIG_ENGINE = "0x97CB9e81d480A2AB03299760654C1DDC0C16bE07";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0x1236010CECea55998384e795B59815D871f5f94d";
var UI_INCENTIVE_DATA_PROVIDER = "0x5d4D4007A4c6336550DdAa2a7c0d5e7972eebd16";
var UI_POOL_DATA_PROVIDER = "0x5598BbFA2f4fE8151f45bBA0a3edE1b54B51a0a9";
var WALLET_BALANCE_PROVIDER = "0x4172E6aAEC070ACB31aaCE343A58c93E4C70f44D";
var WETH_GATEWAY = "0x7B9c12915c594a68dE96201Cbdc79147F09da278";
var CAPS_PLUS_RISK_STEWARD = "0x33AE1f41546a2e05368Bf789b3d868813c0Ae658";
var RISK_STEWARD = "0x677c9f358dA5DC83aF6760a839E4448D72840d04";
var FREEZING_STEWARD = "0x3Ceaf9b6CAb92dFe6302D0CC3F1BA880C28d35e5";
var STATIC_A_TOKEN_FACTORY = "0x02e9b27599C4Bf8f789d34b6E65C51092c3d9FA6";
var REPAY_WITH_COLLATERAL_ADAPTER = "0x86b0521f92a554057e54B93098BA2A6Aaa2F4ACB";
var SWAP_COLLATERAL_ADAPTER = "0x63dfa7c09Dc2Ff4030d6B8Dc2ce6262BF898C8A4";
var DEBT_SWAP_ADAPTER = "0xE28E2c8d240dd5eBd0adcab86fbD79df7a052034";
var WITHDRAW_SWAP_ADAPTER = "0xc0179321f0825c3e0F59Fe7Ca4E40557b97797a3";
var CHAIN_ID = 100;
var ASSETS = {
  WETH: {
    decimals: 18,
    id: 0,
    UNDERLYING: "0x6A023CCd1ff6F2045C3309768eAd9E68F978f6e1",
    A_TOKEN: "0xa818F1B57c201E092C4A2017A91815034326Efd1",
    V_TOKEN: "0x0c0fce05F2314540EcB095bF4D069e5E0ED90fF8",
    INTEREST_RATE_STRATEGY: "0x4cE496f0a390745102540faF041EF92FfD588b44",
    ORACLE: "0xa767f745331D267c7751297D982b050c93985627",
    STATIC_A_TOKEN: "0xD843FB478c5aA9759FeA3f3c98D467e2F136190a"
  },
  wstETH: {
    decimals: 18,
    id: 1,
    UNDERLYING: "0x6C76971f98945AE98dD7d4DFcA8711ebea946eA6",
    A_TOKEN: "0x23e4E76D01B2002BE436CE8d6044b0aA2f68B68a",
    V_TOKEN: "0x9D881f67F20B49243c98f53d2B9E91E39d02Ae09",
    INTEREST_RATE_STRATEGY: "0x4cE496f0a390745102540faF041EF92FfD588b44",
    ORACLE: "0x8Ee42Ba520cA106294163fb8b1ffE9C6Fba35507",
    STATIC_A_TOKEN: "0xECfD0638175e291BA3F784A58FB9D38a25418904"
  },
  GNO: {
    decimals: 18,
    id: 2,
    UNDERLYING: "0x9C58BAcC331c9aa871AFD802DB6379a98e80CEdb",
    A_TOKEN: "0xA1Fa064A85266E2Ca82DEe5C5CcEC84DF445760e",
    V_TOKEN: "0xBc59E99198DbA71985A66E1713cC89FFEC53f7FC",
    INTEREST_RATE_STRATEGY: "0x4cE496f0a390745102540faF041EF92FfD588b44",
    ORACLE: "0x22441d81416430A54336aB28765abd31a792Ad37",
    STATIC_A_TOKEN: "0x2D737e2B0e175f05D0904C208d6C4e40da570f65"
  },
  USDC: {
    decimals: 6,
    id: 3,
    UNDERLYING: "0xDDAfbb505ad214D7b80b1f830fcCc89B60fb7A83",
    A_TOKEN: "0xc6B7AcA6DE8a6044E0e32d0c841a89244A10D284",
    V_TOKEN: "0x5F6f7B0a87CA3CF3d0b431Ae03EF3305180BFf4d",
    INTEREST_RATE_STRATEGY: "0x4cE496f0a390745102540faF041EF92FfD588b44",
    ORACLE: "0x0a2d05bc646C65A029e602c257DfA14adF8BfAd2",
    STATIC_A_TOKEN: "0x270bA1f35D8b87510D24F693fcCc0da02e6E4EeB"
  },
  WXDAI: {
    decimals: 18,
    id: 4,
    UNDERLYING: "0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d",
    A_TOKEN: "0xd0Dd6cEF72143E22cCED4867eb0d5F2328715533",
    V_TOKEN: "0x281963D7471eCdC3A2Bd4503e24e89691cfe420D",
    INTEREST_RATE_STRATEGY: "0x4cE496f0a390745102540faF041EF92FfD588b44",
    ORACLE: "0xE5269eF0CE04E509E8134624c7BF043b21e10897",
    STATIC_A_TOKEN: "0x7f0EAE87Df30C468E0680c83549D0b3DE7664D4B"
  },
  EURe: {
    decimals: 18,
    id: 5,
    UNDERLYING: "0xcB444e90D8198415266c6a2724b7900fb12FC56E",
    A_TOKEN: "0xEdBC7449a9b594CA4E053D9737EC5Dc4CbCcBfb2",
    V_TOKEN: "0xb96404e475f337A7E98e4a541C9b71309BB66c5A",
    INTEREST_RATE_STRATEGY: "0x4cE496f0a390745102540faF041EF92FfD588b44",
    ORACLE: "0xab70BCB260073d036d1660201e9d5405F5829b7a",
    STATIC_A_TOKEN: "0x8418D17640a74F1614AC3E1826F29e78714488a1"
  },
  sDAI: {
    decimals: 18,
    id: 6,
    UNDERLYING: "0xaf204776c7245bF4147c2612BF6e5972Ee483701",
    A_TOKEN: "0x7a5c3860a77a8DC1b225BD46d0fb2ac1C6D191BC",
    V_TOKEN: "0x8Fe06E1D8Aff42Bf6812CacF7854A2249a00bED7",
    INTEREST_RATE_STRATEGY: "0x4cE496f0a390745102540faF041EF92FfD588b44",
    ORACLE: "0x1D0f881Ce1a646E2f27Dec3c57Fa056cB838BCC2",
    STATIC_A_TOKEN: "0xf3f45960f8dE00D8ED614D445a5a268c6F6Dec4f"
  },
  USDCe: {
    decimals: 6,
    id: 7,
    UNDERLYING: "0x2a22f9c3b484c3629090FeED35F17Ff8F88f76F0",
    A_TOKEN: "0xC0333cb85B59a788d8C7CAe5e1Fd6E229A3E5a65",
    V_TOKEN: "0x37B9Ad6b5DC8Ad977AD716e92F49e9D200e58431",
    INTEREST_RATE_STRATEGY: "0x4cE496f0a390745102540faF041EF92FfD588b44",
    ORACLE: "0x0a2d05bc646C65A029e602c257DfA14adF8BfAd2",
    STATIC_A_TOKEN: "0xf0E7eC247b918311afa054E0AEdb99d74c31b809"
  }
};
var E_MODES = {
  "1": {
    label: "ETH correlated",
    collateralBitmap: "3",
    collateralAssets: [
      "0x6A023CCd1ff6F2045C3309768eAd9E68F978f6e1",
      "0x6C76971f98945AE98dD7d4DFcA8711ebea946eA6"
    ],
    borrowableBitmap: "3",
    borrowableAssets: [
      "0x6A023CCd1ff6F2045C3309768eAd9E68F978f6e1",
      "0x6C76971f98945AE98dD7d4DFcA8711ebea946eA6"
    ],
    ltv: 9e3,
    liquidationThreshold: 9300,
    liquidationBonus: 10100
  }
};
var EXTERNAL_LIBRARIES = {};
export {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CAPS_PLUS_RISK_STEWARD,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEBT_SWAP_ADAPTER,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  FREEZING_STEWARD,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  REPAY_WITH_COLLATERAL_ADAPTER,
  RISK_STEWARD,
  STATIC_A_TOKEN_FACTORY,
  SWAP_COLLATERAL_ADAPTER,
  UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY,
  WITHDRAW_SWAP_ADAPTER
};
