// @bgd-labs/aave-address-book/dist/AaveV3BaseSepolia.mjs
// src/ts/AaveV3BaseSepolia.ts
var POOL_ADDRESSES_PROVIDER = "0x150E9a8b83b731B9218a5633F1E804BC82508A46";
var POOL = "0xbE781D7Bdf469f3d94a62Cdcc407aCe106AEcA74";
var POOL_CONFIGURATOR = "0x4200a89Bd036745918889D6dCC5685A4C6F01C24";
var ORACLE = "0x9Ba30437Ba63AA2902319DE1B3f0E25a18826842";
var ACL_ADMIN = "0x6ec33534BE07d45cc4E02Fbd127F8ed2aE919a6b";
var ACL_MANAGER = "0xb880879303D35a82B1888a4A6d47D2ef0653E4A2";
var AAVE_PROTOCOL_DATA_PROVIDER = "0xAF4646B0131af8fc0DC435AF7F7d303Ac131E072";
var POOL_IMPL = "0x07D04EfAAA0Ac69D19d107795aF247C42Eb50F1C";
var POOL_CONFIGURATOR_IMPL = "0x9E7DF170E44093d6738057157CA048794B02555d";
var DEFAULT_INCENTIVES_CONTROLLER = "0x294FF52d234a7e09E9642F846702A45337ceB2E2";
var EMISSION_MANAGER = "0x1F95c29C9E686e4f438ED5D13cF0B7430B9F39ec";
var COLLECTOR = "0x67F521ca716dD9413fd2D2AfdEbEE9285289d2cB";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0xA9E3fFb25C369e44862DD3e87Be4420abb879965";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0x95eeA7A0b16C8ee3A923D3F5ebe6d77C0332084c";
var STATA_FACTORY = "0xAE252DA024783d1813C890d82642bbED120c3093";
var CONFIG_ENGINE = "0x3d2ee1AB8C3a597cDf80273C684dE0036481bE3a";
var L2_ENCODER = "0x0ffE481FBF0AE2282A5E1f701fab266aF487A97D";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0x5A6c2685b9dd22705203C99d7Fc30AE53C4c7513";
var UI_INCENTIVE_DATA_PROVIDER = "0xb0633e01310a09C1Ee71a96c057DcF9c13fd6F62";
var WALLET_BALANCE_PROVIDER = "0xdc5D225Df17df184d11015B91C4A10cd7834e2aC";
var WETH_GATEWAY = "0xd5DDE725b0A2dE43fBDb4E488A7fdA389210d461";
var CHAIN_ID = 84532;
var ASSETS = {
  USDC: {
    decimals: 6,
    id: 0,
    UNDERLYING: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
    A_TOKEN: "0xfE45Bf4dEF7223Ab1Bf83cA17a4462Ef1647F7FF",
    V_TOKEN: "0x5E531B00C86C2D0014020183DaFE7c17C4aA90D8",
    INTEREST_RATE_STRATEGY: "0xff1DB744F1275f8e04A62A7E5D663575F3a774B6",
    ORACLE: "0xd30e2101a97dcbAeBCBC04F14C3f624E67A35165"
  },
  EURC: {
    decimals: 6,
    id: 1,
    UNDERLYING: "0x808456652fdb597867f38412077A9182bf77359F",
    A_TOKEN: "0x70B607b3cdED31635779781d42540580D4Ac18F4",
    V_TOKEN: "0xEa1b67213437E753656B209d1d3d16c1DFE424fd",
    INTEREST_RATE_STRATEGY: "0xff1DB744F1275f8e04A62A7E5D663575F3a774B6",
    ORACLE: "0xd30e2101a97dcbAeBCBC04F14C3f624E67A35165"
  },
  WETH: {
    decimals: 18,
    id: 2,
    UNDERLYING: "0x4200000000000000000000000000000000000006",
    A_TOKEN: "0x6dE9f4b8d4A52D15F1372ef463e27AeAa8a3FdF4",
    V_TOKEN: "0x80bEA6A08B3c2df41B48F27c983C3238f1144093",
    INTEREST_RATE_STRATEGY: "0xff1DB744F1275f8e04A62A7E5D663575F3a774B6",
    ORACLE: "0x4aDC67696bA383F43DD60A9e78F2C97Fbbfc7cb1"
  },
  USDX: {
    decimals: 18,
    id: 3,
    UNDERLYING: "0x3e138010792d63e4af70dD7F2401C9cdE0eaf3C8",
    A_TOKEN: "0x7E60de07156fdA16A74F3700725e2bbB5CFC8CB8",
    V_TOKEN: "0xed47C3087a7B41a4F2cA93747DDDc368f8EcDE47",
    INTEREST_RATE_STRATEGY: "0xff1DB744F1275f8e04A62A7E5D663575F3a774B6",
    ORACLE: "0xd30e2101a97dcbAeBCBC04F14C3f624E67A35165",
    STATA_TOKEN: "0xb17966889E3D914CCf9A11bA0Fd71870B03727AF"
  }
};
var E_MODES = {};
var EXTERNAL_LIBRARIES = {};
export {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  L2_ENCODER,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  STATA_FACTORY,
  UI_INCENTIVE_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY
};
