// @bgd-labs/aave-address-book/dist/AaveV3EthereumEtherFi.mjs
// src/ts/AaveV3EthereumEtherFi.ts
var POOL_ADDRESSES_PROVIDER = "0xeBa440B438Ad808101d1c451C1C5322c90BEFCdA";
var POOL = "0x0AA97c284e98396202b6A04024F5E2c65026F3c0";
var POOL_CONFIGURATOR = "0x8438F4D29D895d75C86BDC25360c25eF0607E65d";
var ORACLE = "0x43b64f28A678944E0655404B0B98E443851cC34F";
var ACL_ADMIN = "0x5300A1a15135EA4dc7aD5a167152C01EFc9b192A";
var ACL_MANAGER = "0x3cE8E2eb6501d4705477643E96881B1bef6A2DB3";
var AAVE_PROTOCOL_DATA_PROVIDER = "0xE7d490885A68f00d9886508DF281D67263ed5758";
var POOL_IMPL = "0x3d881c2Dc90F00e7A52F06155f77FBEC63A779c7";
var POOL_CONFIGURATOR_IMPL = "0x4816b2C2895f97fB918f1aE7Da403750a0eE372e";
var DEFAULT_INCENTIVES_CONTROLLER = "0x8164Cc65827dcFe994AB23944CBC90e0aa80bFcb";
var EMISSION_MANAGER = "0x223d844fc4B006D67c0cDbd39371A9F73f69d974";
var COLLECTOR = "0x464C71f6c2F760DdA6093dCB91C24c39e5d6e18c";
var DEFAULT_A_TOKEN_IMPL_REV_1 = "0xaFFA06528Bd92625de2e7A0cfa0119319265Ea4b";
var DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1 = "0xBb077DaFFeb23B2126E7358b0b122ba6838FB881";
var POOL_ADDRESSES_PROVIDER_REGISTRY = "0xbaA999AC55EAce41CcAE355c77809e68Bb345170";
var REPAY_WITH_COLLATERAL_ADAPTER = "0x23b282c49C88d9161aae14b5eD777B976A5Ae65D";
var SWAP_COLLATERAL_ADAPTER = "0xB04427eFdd15b0EC233400d2F7f7E4fd0291C285";
var WITHDRAW_SWAP_ADAPTER = "0x850347E0cF64fd342A3404c1c5DA21Aa0A46c5c6";
var UI_INCENTIVE_DATA_PROVIDER = "0xe3dFf4052F0bF6134ACb73bEaE8fe2317d71F047";
var UI_POOL_DATA_PROVIDER = "0x3F78BBD206e4D3c504Eb854232EdA7e47E9Fd8FC";
var WALLET_BALANCE_PROVIDER = "0xC7be5307ba715ce89b152f3Df0658295b3dbA8E2";
var WETH_GATEWAY = "0xAB911dFB2bB9e264EE836F30D3367618f8Aef965";
var CONFIG_ENGINE = "0x909bA8DA4c826C62013Ce3A30ce1F42943F3b340";
var STATIC_A_TOKEN_FACTORY = "0x1Ff525426800279843B71C0F818594DeCdC3b522";
var CAPS_PLUS_RISK_STEWARD = "0x1EBdbE77bbDDD284BdCE8D7981D7eD26D6af58cA";
var RISK_STEWARD = "0xBF79d8339303148E345277a994Eb2cD5d82F0067";
var DEBT_SWAP_ADAPTER = "0x40ede12b44d98Fc4E53A2fB027D1BD1846e1690C";
var CHAIN_ID = 1;
var ASSETS = {
  weETH: {
    decimals: 18,
    id: 0,
    UNDERLYING: "0xCd5fE23C85820F7B72D0926FC9b05b43E359b7ee",
    A_TOKEN: "0xbe1F842e7e0afd2c2322aae5d34bA899544b29db",
    V_TOKEN: "0x16264412CB72F0d16A446f7D928Dd0D822810048",
    INTEREST_RATE_STRATEGY: "0xC16666b7FF197427BD255E6961A5F99cfb3A6059",
    ORACLE: "0xf112aF6F0A332B815fbEf3Ff932c057E570b62d3"
  },
  USDC: {
    decimals: 6,
    id: 1,
    UNDERLYING: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    A_TOKEN: "0x7380c583cDe4409eFF5DD3320D93a45D96B80E2e",
    V_TOKEN: "0x9355032d747f1e08F8720CD01950E652eE15cdB7",
    INTEREST_RATE_STRATEGY: "0xC16666b7FF197427BD255E6961A5F99cfb3A6059",
    ORACLE: "0x736bF902680e68989886e9807CD7Db4B3E015d3C"
  },
  PYUSD: {
    decimals: 6,
    id: 2,
    UNDERLYING: "0x6c3ea9036406852006290770BEdFcAbA0e23A0e8",
    A_TOKEN: "0xdF7f48892244C6106EA784609f7de10AB36F9c7e",
    V_TOKEN: "0xD2cf07dEE40d3D530D15b88d689f5cd97A31FC3D",
    INTEREST_RATE_STRATEGY: "0xC16666b7FF197427BD255E6961A5F99cfb3A6059",
    ORACLE: "0x150bAe7Ce224555D39AfdBc6Cb4B8204E594E022"
  },
  FRAX: {
    decimals: 18,
    id: 3,
    UNDERLYING: "0x853d955aCEf822Db058eb8505911ED77F175b99e",
    A_TOKEN: "0x6914ECCf50837dC61b43ee478a9BD9B439648956",
    V_TOKEN: "0xfd3aDA5AAbdc6531C7C2AC46c00eBf870f5a0E6B",
    INTEREST_RATE_STRATEGY: "0xC16666b7FF197427BD255E6961A5F99cfb3A6059",
    ORACLE: "0x45D270263BBee500CF8adcf2AbC0aC227097b036"
  }
};
var E_MODES = {};
var EXTERNAL_LIBRARIES = {};
export {
  AAVE_PROTOCOL_DATA_PROVIDER,
  ACL_ADMIN,
  ACL_MANAGER,
  ASSETS,
  CAPS_PLUS_RISK_STEWARD,
  CHAIN_ID,
  COLLECTOR,
  CONFIG_ENGINE,
  DEBT_SWAP_ADAPTER,
  DEFAULT_A_TOKEN_IMPL_REV_1,
  DEFAULT_INCENTIVES_CONTROLLER,
  DEFAULT_VARIABLE_DEBT_TOKEN_IMPL_REV_1,
  EMISSION_MANAGER,
  EXTERNAL_LIBRARIES,
  E_MODES,
  ORACLE,
  POOL,
  POOL_ADDRESSES_PROVIDER,
  POOL_ADDRESSES_PROVIDER_REGISTRY,
  POOL_CONFIGURATOR,
  POOL_CONFIGURATOR_IMPL,
  POOL_IMPL,
  REPAY_WITH_COLLATERAL_ADAPTER,
  RISK_STEWARD,
  STATIC_A_TOKEN_FACTORY,
  SWAP_COLLATERAL_ADAPTER,
  UI_INCENTIVE_DATA_PROVIDER,
  UI_POOL_DATA_PROVIDER,
  WALLET_BALANCE_PROVIDER,
  WETH_GATEWAY,
  WITHDRAW_SWAP_ADAPTER
};
