// @bgd-labs/aave-address-book/dist/MiscScroll.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ts/MiscScroll.ts
var MiscScroll_exports = {};
__export(MiscScroll_exports, {
  CHAIN_ID: () => CHAIN_ID,
  CREATE_3_FACTORY: () => CREATE_3_FACTORY,
  PROTOCOL_GUARDIAN: () => PROTOCOL_GUARDIAN,
  PROXY_ADMIN: () => PROXY_ADMIN,
  TRANSPARENT_PROXY_FACTORY: () => TRANSPARENT_PROXY_FACTORY,
  wstETH_stETH_AGGREGATOR: () => wstETH_stETH_AGGREGATOR
});
module.exports = __toCommonJS(MiscScroll_exports);
var TRANSPARENT_PROXY_FACTORY = "0x3cd1dFB81C50A5300C60a181ED145a7286d81e0a";
var PROXY_ADMIN = "0x782559e349b084bB7C07c08404aE6E3436cDAE2E";
var CREATE_3_FACTORY = "0xC6f44D0daC966658Db36742f5153bDa0EBbb45ab";
var PROTOCOL_GUARDIAN = "0xCb45E82419baeBCC9bA8b1e5c7858e48A3B26Ea6";
var wstETH_stETH_AGGREGATOR = "0xE61Da4C909F7d86797a0D06Db63c34f76c9bCBDC";
var CHAIN_ID = 534352;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CHAIN_ID,
  CREATE_3_FACTORY,
  PROTOCOL_GUARDIAN,
  PROXY_ADMIN,
  TRANSPARENT_PROXY_FACTORY,
  wstETH_stETH_AGGREGATOR
});
