// @bgd-labs/aave-address-book/dist/MiscBase.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ts/MiscBase.ts
var MiscBase_exports = {};
__export(MiscBase_exports, {
  AAVE_CL_ROBOT_OPERATOR: () => AAVE_CL_ROBOT_OPERATOR,
  CHAIN_ID: () => CHAIN_ID,
  CREATE_3_FACTORY: () => CREATE_3_FACTORY,
  LEGACY_BRIDGE_EXECUTOR: () => LEGACY_BRIDGE_EXECUTOR,
  PARASWAP_FEE_CLAIMER: () => PARASWAP_FEE_CLAIMER,
  PROTOCOL_GUARDIAN: () => PROTOCOL_GUARDIAN,
  PROXY_ADMIN: () => PROXY_ADMIN,
  TRANSPARENT_PROXY_FACTORY: () => TRANSPARENT_PROXY_FACTORY,
  cbETH_ETH_AGGREGATOR: () => cbETH_ETH_AGGREGATOR,
  wstETH_stETH_AGGREGATOR: () => wstETH_stETH_AGGREGATOR
});
module.exports = __toCommonJS(MiscBase_exports);
var PARASWAP_FEE_CLAIMER = "0xAe940e61E9863178b71500c9B5faE2a04Da361a1";
var TRANSPARENT_PROXY_FACTORY = "0xd9Ca4878dd38B021583c1B669905592EAe76E044";
var PROXY_ADMIN = "0xc85b1E333aecc99340b2320493Fe2d22b8734795";
var CREATE_3_FACTORY = "0x3b56998Ec06477704622ca8e2eA1b4db134cec32";
var AAVE_CL_ROBOT_OPERATOR = "0x88db99eeBb390a2a4DcAC2E1DDb09c07E911C5C3";
var PROTOCOL_GUARDIAN = "0x56C1a4b54921DEA9A344967a8693C7E661D72968";
var wstETH_stETH_AGGREGATOR = "0xB88BAc61a4Ca37C43a3725912B1f472c9A5bc061";
var cbETH_ETH_AGGREGATOR = "0x868a501e68F3D1E89CfC0D22F6b22E8dabce5F04";
var LEGACY_BRIDGE_EXECUTOR = "0xA9F30e6ED4098e9439B2ac8aEA2d3fc26BcEbb45";
var CHAIN_ID = 8453;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AAVE_CL_ROBOT_OPERATOR,
  CHAIN_ID,
  CREATE_3_FACTORY,
  LEGACY_BRIDGE_EXECUTOR,
  PARASWAP_FEE_CLAIMER,
  PROTOCOL_GUARDIAN,
  PROXY_ADMIN,
  TRANSPARENT_PROXY_FACTORY,
  cbETH_ETH_AGGREGATOR,
  wstETH_stETH_AGGREGATOR
});
