// @bgd-labs/aave-address-book/dist/MiscOptimism.mjs
// src/ts/MiscOptimism.ts
var PARASWAP_FEE_CLAIMER = "0x9abf798f5314BFd793A9E57A654BEd35af4A1D60";
var TRANSPARENT_PROXY_FACTORY = "0x984B710D22730f799312513A10C1382e9D1Fa689";
var PROXY_ADMIN = "0xD3cF979e676265e4f6379749DECe4708B9A22476";
var CREATE_3_FACTORY = "0x3b56998Ec06477704622ca8e2eA1b4db134cec32";
var AAVE_CL_ROBOT_OPERATOR = "0x55Cf9583D7D30DC4936bAee1f747591dBECe5df7";
var PROTOCOL_GUARDIAN = "0x56C1a4b54921DEA9A344967a8693C7E661D72968";
var AAVE_MERKLE_DISTRIBUTOR = "0x1685D81212580DD4cDA287616C2f6F4794927e18";
var wstETH_stETH_AGGREGATOR = "0xe59EBa0D492cA53C6f46015EEa00517F2707dc77";
var rETH_ETH_AGGREGATOR = "0x22F3727be377781d1579B7C9222382b21c9d1a8f";
var LEGACY_BRIDGE_EXECUTOR = "0x7d9103572bE58FfE99dc390E8246f02dcAe6f611";
var CHAIN_ID = 10;
export {
  AAVE_CL_ROBOT_OPERATOR,
  AAVE_MERKLE_DISTRIBUTOR,
  CHAIN_ID,
  CREATE_3_FACTORY,
  LEGACY_BRIDGE_EXECUTOR,
  PARASWAP_FEE_CLAIMER,
  PROTOCOL_GUARDIAN,
  PROXY_ADMIN,
  TRANSPARENT_PROXY_FACTORY,
  rETH_ETH_AGGREGATOR,
  wstETH_stETH_AGGREGATOR
};
