// @bgd-labs/aave-address-book/dist/MiscArbitrum.mjs
// src/ts/MiscArbitrum.ts
var PARASWAP_FEE_CLAIMER = "0x9abf798f5314BFd793A9E57A654BEd35af4A1D60";
var TRANSPARENT_PROXY_FACTORY = "0x9B643d4aC937BEE8Bfb1Ec76401C22E18C318429";
var PROXY_ADMIN = "0xD3cF979e676265e4f6379749DECe4708B9A22476";
var CREATE_3_FACTORY = "0x3b56998Ec06477704622ca8e2eA1b4db134cec32";
var AAVE_CL_ROBOT_OPERATOR = "0xaa944aD95e51CB83C1f35FAEEDfC7d2c31B0BB4d";
var PROTOCOL_GUARDIAN = "0xCb45E82419baeBCC9bA8b1e5c7858e48A3B26Ea6";
var wstETH_stETH_AGGREGATOR = "0xB1552C5e96B312d0Bf8b554186F846C40614a540";
var rETH_ETH_AGGREGATOR = "0xF3272CAfe65b190e76caAF483db13424a3e23dD2";
var GHO_CCIP_TOKEN_POOL = "0xF168B83598516A532a85995b52504a2Fa058C068";
var LEGACY_BRIDGE_EXECUTOR = "0x7d9103572bE58FfE99dc390E8246f02dcAe6f611";
var CHAIN_ID = 42161;
export {
  AAVE_CL_ROBOT_OPERATOR,
  CHAIN_ID,
  CREATE_3_FACTORY,
  GHO_CCIP_TOKEN_POOL,
  LEGACY_BRIDGE_EXECUTOR,
  PARASWAP_FEE_CLAIMER,
  PROTOCOL_GUARDIAN,
  PROXY_ADMIN,
  TRANSPARENT_PROXY_FACTORY,
  rETH_ETH_AGGREGATOR,
  wstETH_stETH_AGGREGATOR
};
