===================================================================
--- src/contracts/protocol/pool/L2Pool.sol
+++ lib/aave-v3-origin-private/src/contracts/protocol/pool/L2Pool.sol
@@ -16,9 +16,9 @@
     function supply(bytes32 args) external override {
         (address asset, uint256 amount, uint16 referralCode) = CalldataLogic
             .decodeSupplyParams(_reservesList, args);
 
-        supply(asset, amount, msg.sender, referralCode);
+        supply(asset, amount, _msgSender(), referralCode);
     }
 
     /// @inheritdoc IL2Pool
     function supplyWithPermit(
@@ -36,9 +36,9 @@
 
         supplyWithPermit(
             asset,
             amount,
-            msg.sender,
+            _msgSender(),
             referralCode,
             deadline,
             v,
             r,
@@ -52,9 +52,9 @@
             _reservesList,
             args
         );
 
-        return withdraw(asset, amount, msg.sender);
+        return withdraw(asset, amount, _msgSender());
     }
 
     /// @inheritdoc IL2Pool
     function borrow(bytes32 args) external override {
@@ -64,9 +64,9 @@
             uint256 interestRateMode,
             uint16 referralCode
         ) = CalldataLogic.decodeBorrowParams(_reservesList, args);
 
-        borrow(asset, amount, interestRateMode, referralCode, msg.sender);
+        borrow(asset, amount, interestRateMode, referralCode, _msgSender());
     }
 
     /// @inheritdoc IL2Pool
     function repay(bytes32 args) external override returns (uint256) {
@@ -75,9 +75,9 @@
             uint256 amount,
             uint256 interestRateMode
         ) = CalldataLogic.decodeRepayParams(_reservesList, args);
 
-        return repay(asset, amount, interestRateMode, msg.sender);
+        return repay(asset, amount, interestRateMode, _msgSender());
     }
 
     /// @inheritdoc IL2Pool
     function repayWithPermit(
@@ -97,9 +97,9 @@
             repayWithPermit(
                 asset,
                 amount,
                 interestRateMode,
-                msg.sender,
+                _msgSender(),
                 deadline,
                 v,
                 r,
                 s
